enterprise/backend/src/metabase_enterprise/sso/integrations/sso_utils.clj

```diff
@@ -36,13 +36,15 @@
 
 (defn fetch-and-update-login-attributes!
   "Update the login attributes for the user at `email`. This call is a no-op if the login attributes are the same"
-  [email new-user-attributes]
-  (when-let [{:keys [id login_attributes] :as user} (db/select-one User :%lower.email (u/lower-case-en email))]
-    (if (= login_attributes new-user-attributes)
-      user
-      (do
-        (db/update! User id :login_attributes new-user-attributes)
-        (User id)))))
+  [{:keys [email] :as user-attributes}]
+  (when-let [{:keys [id] :as user} (db/select-one User :%lower.email (u/lower-case-en email))]
+    (let [attribute-keys (keys user-attributes)
+          user-attributes (into {} (filter second user-attributes))] ;; remove attributes with `nil` values
```

Comment:
I think essentially the same logic is already implemented for LDAP in `metabase-enterprise.enhancements.integrations.ldap/attribute-synced-user`, though the code looks very different. Any chance we could have a single implementation for this, shared by LDAP, JWT and SAML? (I don't really care which implementation you choose, but it would be nice to not have the duplication if possible. And if you can verify that the behavior for LDAP doesn't change.)

Reply:
I think LDAP may benefit from a more thorough cleanup to be honest, but I've now altered things a bit to use the `sso-utils/fetch-and-update-user!` function, so we're getting closer to this unification :)

Reply:
I've actually reverted the LDAP change. I got stuck and want to move this PR forward, so I'll address LDAP in a follow on PR.